### PR TITLE
ENH: allow gzip de-compression for files specified by a url

### PR DESCRIPTION
**Pull request by mdagost**
Address #8685 .  The file in the test `test_url_gz` currently points to my fork to demonstrate that the tests pass, but should be changed to point to the real pydata master after merging.

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -135,7 +142,7 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None):
 
     if _is_url(filepath_or_buffer):
         req = _urlopen(str(filepath_or_buffer))
```
Comment (jseabold) [inline pandas/io/common.py]: You can also do `compression = req.headers['Content-Encoding']` here, if it's being served as a gzip encoded.

Comment (mdagost) [inline pandas/io/common.py]: Yeah, that's a great idea.  I couldn't set a Content-Encoding header on a file in github, but I can in s3, so I'll add that code in and a new test.

**Comment by mdagost** [comment]
This works for `compression="infer"` now as well by looking at the HTTP `Content-Encoding` header like @jseabold suggested.  The test `test_url_gz_infer` uses a file hosted on an s3 bucket that I own where I set it to return `Content-Encoding=gzip`.  Assuming this gets merged, we'd want to move that into the s3 bucket that pandas uses elsewhere in the tests.

**Comment by mdagost** [comment]
Any admins have time to look at this PR?

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -234,8 +235,11 @@ def _read(filepath_or_buffer, kwds):
     if skipfooter is not None:
         kwds['skip_footer'] = skipfooter
 
-    filepath_or_buffer, _ = get_filepath_or_buffer(filepath_or_buffer,
-                                                   encoding)
+    compression = kwds.get('compression', None)
```
Comment (jreback) [inline pandas/io/parsers.py]: put this directly in the call to ``get_filepath_or_buffer(....., compression=kws.get('compression', None)``

File: pandas/io/tests/test_parsers.py [hunk 3]
```diff
@@ -3070,6 +3070,26 @@ def test_whitespace_lines(self):
         df = self.read_csv(StringIO(data))
         tm.assert_almost_equal(df.values, expected)
 
+    @tm.network
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: shouldn't this test be the same as the one below (e.g. with the same URL), except for the difference in the compression arg?

File: pandas/io/tests/test_parsers.py [hunk 4]
```diff
@@ -3070,6 +3070,26 @@ def test_whitespace_lines(self):
         df = self.read_csv(StringIO(data))
         tm.assert_almost_equal(df.values, expected)
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: I would put both of these in a new ``TestNetwork`` (below TestS3), or just throw them in there ok too. You can have some of the boilerplate in the ``setUp`` creation call. Eg. the filepath to salary.

**Comment by jreback** [comment]
at the end pls squash to a single commit. pls add a release note in whatsnew/0.17.0.

Here are the contribution guidelines for your [reference](http://pandas.pydata.org/pandas-docs/stable/contributing.html)

File: pandas/io/tests/test_parsers.py (hunk 3 above)
Comment (mdagost) [inline pandas/io/tests/test_parsers.py]: With `compression="infer"`, we look at the `Content-Encoding` header from the response.  There's no way to get `raw.github` to reply with a `Content-Encoding` header, so I had to use an s3 url, where you can set arbitrary response headers.

**Comment by mdagost** [comment]
@jreback Once the tests all pass, I'll squash into a single commit.  There's still the issue of changing the test url's to point to the official github repo and s3 bucket.  Not sure the best way to handle that.

**Comment by mdagost** [comment]
@jreback Okay.  This should be good to go.  Thanks for the comments!

**Comment by jreback** [comment]
I pushed the .gz file, so you can use this as a stable address: https://github.com/pydata/pandas/blob/master/pandas/io/tests/data/salary.table.gz

pls change, squash, and ping when green

**Comment by mdagost** [comment]
@jreback Thanks!  What about the s3 file?  Are you able to post that too and set s3 to return the Content-Encoding gzip header?

**Comment by jreback** [comment]
so is a that a public bucket you setup? any problem leaving it for now?

**Comment by mdagost** [comment]
Yep.  It's public.  I've got no problem leaving it.  Just didn't know if you guys wanted to have it look more "official."  I'll make that other change to the github link.

**Comment by jreback** [comment]
we don't have an official s3 bucket for this ATM. But I think I can get access to one where we can do stuff like this.

**Comment by mdagost** [comment]
Ah.  Okay.  I saw `s3://nyqpug/` in the tests and assumed that was official.  Happy to leave mine in there for awhile until you figure it out.

**Comment by jreback** [comment]
hmm, that is some old stuff IIRC. Wll try to set this up for later today.

**Comment by jreback** [comment]
ok, going to merge this then fix it with a new bucket. thanks!

**Comment by mdagost** [comment]
Oh shoot.  I didn't get a chance to change the regular .gz github link.

**Comment by jreback** [comment]
no worries, going to merge this shortly: https://github.com/jreback/pandas/commit/2f0c344191be435b0302ad8e8af4ca3183f2fcf5

This is a public bucket for pandas (though read-only)

**Pull request merged**